### ISSUE TITLE
Object Detection training purely on Colab

### ISSUE THREAD

**Issue by khanhlvg** [ml-gde/e2e-tflite-tutorials#22]
Hi @sayakpaul , could you try to put together a notebook that show how to train a TF2 Object Detection model using Colab GPU?

@farmaker47 FYI

**Comment by sayakpaul** [comment]
@khanhlvg thanks for creating the issue. A few things:

- This [Colab Notebook](https://github.com/tensorflow/models/blob/master/research/object_detection/colab_tutorials/eager_few_shot_od_training_tf2_colab.ipynb) shows few-shot learning in the context of object detection and it includes training too. 
- Should I show the classic declarative configuration based training on Colab i.e. the non eager mode? For more context, this is what I did in [here](https://github.com/sayakpaul/E2E-Object-Detection-in-TFLite/blob/master/Training_a_pets_detector_model_within_minutes_with_TFOD_API.ipynb) but on Cloud TPUs.

**Comment by khanhlvg** [comment]
I think using the non-eager way with the existing [script](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_training_and_evaluation.md#local) is fine as the goal is to provide the most simple e2e flow that run on Colab so that anyone can try it out without any cost.

I believe you can reuse most of your Cloud TPUs code, except running them on Colab rather than starting a Cloud AI Platform job.

**Comment by sayakpaul** [comment]
Fair enough!

**Comment by sayakpaul** [comment]
@khanhlvg I have been able to kickstart the training process in Colab (I have kept the number of steps to 2000 intentionally because of the runtime). Here's my [Colab Notebook](https://colab.research.google.com/drive/1Czj3oQjWWKyxa_QtJFHDtXdNnzFMyjgk?usp=sharing).

I used MobileDet (TensorFlow 1) for the purpose. Here's the ToC - 

![image](https://user-images.githubusercontent.com/22957388/93008902-db84c600-f597-11ea-9118-b4448a7df357.png)

I will update here when it's done.

**Comment by sayakpaul** [comment]
@khanhlvg here's the updated [Colab Notebook](https://colab.research.google.com/github/sayakpaul/E2E-Object-Detection-in-TFLite/blob/master/Training_MobileDet_Custom_Dataset.ipynb). Let me know what you think.

**Comment by khanhlvg** [comment]
What was the mAP you got for the pet dataset when training on Cloud TPU for long enough? How many training steps require to reach that mAP level?

If I remember correctly, the pet dataset was very big (several GBs) so 2000 steps with batch size = 32, as you pointed out, was not enough to train a pet detector with good accuracy.

If that's the case, I think there's 2 approaches:

1. Choose a smaller object detection dataset to demonstrate proper training on Colab.
2. Just focus on training with AI Platform and TPUs (which you already did)

WDYT?

**Comment by sayakpaul** [comment]
> What was the mAP you got for the pet dataset when training on Cloud TPU for long enough?

I don't exactly remember but it was good enough (with 20 minutes of TPU training). The batch size was 512 and the number of steps was 2000 (here's the [configuration file](https://gist.github.com/sayakpaul/70c414fbe581def12f4b500659372c4f)). 

I am explicitly mentioning the accuracy part being low due to the constraints. Practically speaking, for getting a reasonable detection performance at least an hour of GPU training is needed. For an extremely small number of samples, one could go for few-shot learning I would think. 

> Choose a smaller object detection dataset to demonstrate proper training on Colab.

I think you probably meant the accuracy part because even if the dataset was small the training pipeline wouldn't really change. nor the hyperparameters (yes, I tuned other hyperparameters too such as `warmup_steps`). I would probably train it longer, but that should be it.

If you have a smaller dataset in mind already please feel free to pass that along. But note that it would require additional data preprocessing steps that might make the tutorial more involved and it might break the linear flow of the notebook.

**Comment by khanhlvg** [comment]
> I think you probably meant the accuracy part because even if the dataset was small the training pipeline wouldn't really change. nor the hyperparameters (yes, I tuned other hyperparameters too such as warmup_steps). I would probably train it longer, but that should be it.

Yes I mean with a smaller dataset, the model can reach an acceptable accuracy within a more reasonable time on Colab GPU. 

> If you have a smaller dataset in mind already please feel free to pass that along, I can take a look.

I searched on Kaggle and found this [fruit dataset](https://www.kaggle.com/mbkinaci/fruit-images-for-object-detection) which might be interesting.

> But note that it would require additional data preprocessing steps that might make the tutorial more involved and it might break the linear flow of the notebook. With reference to the initial objective, this may not be ideal.

That's true. But I think in a practical use case, developer will need to create their own training dataset with some tool such as [labelImg](https://github.com/tzutalin/labelImg). Therefore, I think teaching how to prepare data for training should be an important step.

**Comment by sayakpaul** [comment]
> That's true. But I think in a practical use case, developer will need to create their own training dataset with some tool such as labelImg. Therefore, I think teaching how to prepare data for training should be an important step.

I couldn't agree more. At work, we have used labelImg along with other tools like [Explosion](https://prodi.gy/features/computer-vision). But it only addresses one aspect of the pipeline (it's the **annotation** part). To be able to train with the TFOD API you would need to pack everything in TFRecords (at least if you are using it in a non-eager mode). 

I will take a look at the dataset you mentioned.

**Comment by sayakpaul** [comment]
@khanhlvg ready: https://github.com/sayakpaul/E2E-Object-Detection-in-TFLite/tree/master/colab_training.

**Comment by khanhlvg** [comment]
Wow this is so quick! Thanks Sayak.

Did you find the accuracy better with a smaller dataset (the fruit dataset)?

**Comment by sayakpaul** [comment]
Here's the **mAP@.50IOU** ( standard in object detection evaluation)

![image](https://user-images.githubusercontent.com/22957388/93178783-58f43600-f752-11ea-8667-b9bff490ad68.png)

I think it's pretty good.

**Comment by sayakpaul** [comment]
@khanhlvg moving this to done. Feel free to open it.
